File: articles/calculus-for-systems-modeling/articles/stiff-systems-and-computational-difficulty/fortran/stiff_systems_difficulty.f90
program stiff_systems_difficulty
  implicit none
  integer :: i
  real(8), dimension(4) :: hs
  real(8) :: y0, lambda, stop_time, exact_final, explicit_final, implicit_final, h, eamp, iamp

  y0 = 1.0d0
  lambda = -50.0d0
  stop_time = 1.0d0
  hs = (/0.1d0, 0.05d0, 0.025d0, 0.01d0/)
  exact_final = y0 * exp(lambda * stop_time)

  print '(A)', 'step_size eigenvalue method amplification_factor final_value exact_final_value absolute_error'
  do i = 1, 4
    h = hs(i)
    explicit_final = explicit_value(y0, lambda, h, stop_time)
    implicit_final = implicit_value(y0, lambda, h, stop_time)
    eamp = abs(1.0d0 + h * lambda)
    iamp = abs(1.0d0 / (1.0d0 - h * lambda))
    print '(F10.4,F12.4,A18,4F18.10)', h, lambda, ' explicit_euler ', eamp, explicit_final, exact_final, abs(explicit_final - exact_final)
    print '(F10.4,F12.4,A18,4F18.10)', h, lambda, ' implicit_euler ', iamp, implicit_final, exact_final, abs(implicit_final - exact_final)
  end do

contains
  real(8) function explicit_value(y0, lambda, h, stop_time)
    real(8), intent(in) :: y0, lambda, h, stop_time
    integer :: step, steps
    real(8) :: y, amp
    steps = nint(stop_time / h)
    amp = 1.0d0 + h * lambda
    y = y0
    do step = 1, steps
      y = amp * y
    end do
    explicit_value = y
  end function

  real(8) function implicit_value(y0, lambda, h, stop_time)
    real(8), intent(in) :: y0, lambda, h, stop_time
    integer :: step, steps
    real(8) :: y, amp
    steps = nint(stop_time / h)
    amp = 1.0d0 / (1.0d0 - h * lambda)
    y = y0
    do step = 1, steps
      y = amp * y
    end do
    implicit_value = y
  end function
end program

File: articles/calculus-for-systems-modeling/articles/stiff-systems-and-computational-difficulty/c/stiff_systems_difficulty.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

static double exact_solution(double t, double y0, double lambda){ return y0 * exp(lambda * t); }
static double explicit_value(double y0, double lambda, double h, double stop_time){
  int steps = (int)round(stop_time / h);
  double amp = 1.0 + h * lambda;
  double y = y0;
  for(int i=0; i<steps; i++){ y *= amp; }
  return y;
}
static double implicit_value(double y0, double lambda, double h, double stop_time){
  int steps = (int)round(stop_time / h);
  double amp = 1.0 / (1.0 - h * lambda);
  double y = y0;
  for(int i=0; i<steps; i++){ y *= amp; }
  return y;
}
int main(void){
  const double y0 = 1.0, lambda = -50.0, stop_time = 1.0;
  const double exact_final = exact_solution(stop_time, y0, lambda);
  const double hs[4] = {0.1, 0.05, 0.025, 0.01};
  printf("step_size,eigenvalue,method,amplification_factor,stability_status,final_value,exact_final_value,absolute_error,warning\n");
  for(int i=0; i<4; i++){
    double h = hs[i];
    double ev = explicit_value(y0, lambda, h, stop_time);
    double eamp = fabs(1.0 + h * lambda);
    double iv = implicit_value(y0, lambda, h, stop_time);
    double iamp = fabs(1.0 / (1.0 - h * lambda));
    printf("%.6f,%.6f,explicit_euler,%.12f,%s,%.12f,%.12f,%.12f,Explicit methods may require very small steps on stiff systems.\n", h, lambda, eamp, eamp <= 1.0 ? "stable_for_test_problem" : "unstable_for_test_problem", ev, exact_final, fabs(ev - exact_final));
    printf("%.6f,%.6f,implicit_euler,%.12f,%s,%.12f,%.12f,%.12f,Implicit stability does not remove accuracy review.\n", h, lambda, iamp, iamp <= 1.0 ? "stable_for_test_problem" : "unstable_for_test_problem", iv, exact_final, fabs(iv - exact_final));
  }
  return EXIT_SUCCESS;
}
